src/psyclone/domain/lfric/lfric_types.py

```diff
@@ -48,206 +49,298 @@ class LFRicTypes:
+        that the static internal dictionary is initialised (so it acts like
+        a singleton in that regard). Then it will return the value the user
+        asked for.
 
-        :returns: the singleton instance of LFRicTypes.
-        :rtype: :py:class:`psyclone.domain.lfric.LFRicTypes`
+        :param str name: the name to query for.
+
+        :returns: the corresponding object, which can be a class or an \
+            instance.
+        :rtype: object (various types)
 
         '''
-        if LFRicTypes._instance is None:
-            # Return a new instance. The constructor will set _instance
-            # in this case
-            return super().__new__(cls)
+        if not LFRicTypes._name_to_class:
+            LFRicTypes.init()
 
-        # Return the existing instance, in which case the constructor will
-        # not re-initialise the internal data structures
-        return LFRicTypes._instance
+        value = LFRicTypes._name_to_class.get(name, None)
+        if value:
+            return value
 
-    # ------------------------------------------------------------------------
-    def __init__(self):
+        # Unknown keyword
+        raise InternalError(f"Unknown type '{name}'. Valid values are "
+                            f"{LFRicTypes._name_to_class.keys()}")
 
-        # Test if this is returning the existing instance, if so, skip
-        # initialisation
-        if self == LFRicTypes._instance:
-            return
+    # ------------------------------------------------------------------------
+    def __call__(self):
+        '''This function is only here to trick pylint into thinking that
+        the object returned from __new__ is callable, meaning that code like:
+        ``LFRicTypes("LFRicIntegerScalarDataType")()`` does not trigger
+        a pyling warning about not being callable.
+        '''
 
-        # First time __init__ is called, initialise all data structures.
-        LFRicTypes._instance = self
+    # ------------------------------------------------------------------------
+    @staticmethod
+    def init():
+        '''This method constructs the required classes and instances, and
+        puts them into the dictionary.
+        '''
 
-        self._name_to_class = {}
+        # The global mapping of names to the corresponding classes or instances
+        LFRicTypes._name_to_class = {}
 
-        self._create_modules()
-        self._create_generic_scalars()
-        self._create_lfric_dimension()
-        self._create_specific_scalars()
-        self._create_fields()
+        LFRicTypes._create_precision_from_const_module()
+        LFRicTypes._create_generic_scalars()
+        LFRicTypes._create_lfric_dimension()
+        LFRicTypes._create_specific_scalars()
+        LFRicTypes._create_fields()
         # Generate LFRic vector-field-data symbols as subclasses of
         # field-data symbols
         for intrinsic in ["Real", "Integer", "Logical"]:
```

Comment:
Can you use `LFRicConstants.VALID_FIELD_INTRINSIC_TYPES` here instead? (I think there's a `str` method to capitalise the first char?)

Reply:
Yes, done - I assume you meant to then add 'logical' to VALID_FIELD_INTRINSIC_TYPE, where it wasn't before.

Reply:
I'm not sure I did mean that. There doesn't appear to be support for logical fields in LFRic infrastructure so we shouldn't support them here either.